frontend/javascripts/dashboard/dataset/dataset_collection_context.tsx

```diff
@@ -0,0 +1,276 @@
+import React, { createContext, useCallback, useContext, useEffect, useMemo, useState } from "react";
+import type { DatasetFilteringMode } from "dashboard/dataset_view";
+import type { APIMaybeUnimportedDataset, APIDatasetId, APIDataset } from "types/api_flow_types";
+import { getDatastores, triggerDatasetCheck } from "admin/admin_rest_api";
+import UserLocalStorage from "libs/user_local_storage";
+import _ from "lodash";
+import {
+  useFolderTreeQuery,
+  useDatasetsInFolderQuery,
+  useDatasetSearchQuery,
+  useCreateFolderMutation,
+  useUpdateFolderMutation,
+  useMoveFolderMutation,
+  useDeleteFolderMutation,
+  useUpdateDatasetMutation,
+  useFolderQuery,
+} from "./queries";
+import { useIsMutating } from "@tanstack/react-query";
+
+type Options = {
+  datasetFilteringMode?: DatasetFilteringMode;
+  applyUpdatePredicate?: (datasets: Array<APIMaybeUnimportedDataset>) => boolean;
+  isCalledFromCheckDatasets?: boolean;
+};
+export type DatasetCollectionContextValue = {
+  datasets: Array<APIMaybeUnimportedDataset>;
+  isLoading: boolean;
+  isChecking: boolean;
+  checkDatasets: () => Promise<void>;
+  fetchDatasets: (options?: Options) => Promise<void>;
+  reloadDataset: (
+    datasetId: APIDatasetId,
+    datasetsToUpdate?: Array<APIMaybeUnimportedDataset>,
+  ) => Promise<void>;
+  updateCachedDataset: (dataset: APIDataset) => Promise<void>;
+  activeFolderId: string | null;
+  setActiveFolderId: (id: string) => void;
+  supportsFolders: true;
+  globalSearchQuery: string | null;
+  setGlobalSearchQuery: (val: string | null) => void;
+  queries: {
+    folderTreeQuery: ReturnType<typeof useFolderTreeQuery>;
+    datasetsInFolderQuery: ReturnType<typeof useDatasetsInFolderQuery>;
+    datasetSearchQuery: ReturnType<typeof useDatasetSearchQuery>;
+    createFolderMutation: ReturnType<typeof useCreateFolderMutation>;
+    updateFolderMutation: ReturnType<typeof useUpdateFolderMutation>;
+    moveFolderMutation: ReturnType<typeof useMoveFolderMutation>;
+    deleteFolderMutation: ReturnType<typeof useDeleteFolderMutation>;
+    updateDatasetMutation: ReturnType<typeof useUpdateDatasetMutation>;
+  };
+};
+
+export const DatasetCollectionContext = createContext<DatasetCollectionContextValue | undefined>(
+  undefined,
+);
+
+export const useDatasetCollectionContext = () => {
+  const context = useContext(DatasetCollectionContext);
+  if (!context)
+    throw new Error(
+      "No DatasetCollectionContext.Provider found when calling useDatasetCollectionContext.",
+    );
+  return context;
+};
+
+const ACTIVE_FOLDER_ID_STORAGE_KEY = "activeFolderId";
+
+export default function DatasetCollectionContextProvider({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  const [activeFolderId, setActiveFolderId] = useState<string | null>(
+    UserLocalStorage.getItem(ACTIVE_FOLDER_ID_STORAGE_KEY) || null,
```

Comment:
Hmm, good question. No strong opinion, but I would probably expect it to be at the root folder when I open it again. That seems to be how google tools behave as well. On the other hand, if a user works with the same three datasets in a particular folder for an extended period of time, they might find the additional clicks bothersome. I’d say it depends on how quickly they can get to their folder again … (assuming they don’t want to bookmark the url with folder id). Is the left tree view expanded by default? If not, maybe its last expansion state could be persisted? Or alternatively, we could add a “bookmarks” section or even “recently used“ for folders and/or datasets? Those would be follow-up issues of course.

Reply:
> On the other hand, if a user works with the same three datasets in a particular folder for an extended period of time, they might find the additional clicks bothersome.

That's what I had in mind.

> Or alternatively, we could add a “bookmarks” section or even “recently used“ for folders and/or datasets?

Yes, I think this could be a cool upgrade in the future. However, let's wait for some real user feedback.

I'd vote to keep it now as is (i.e., remember the last active folder).